code/parse/sexp.h

```diff
@@ -1142,6 +1181,59 @@ typedef struct sexp_variable {
 	char	variable_name[TOKEN_LENGTH];
 } sexp_variable;
 
+struct sexp_container
+{
+	// meta-character for containers in text replacement, etc.
+	static constexpr char DELIM = '&';
+	// applies to list data, map keys, and map data
+	static constexpr int VALUE_MAX_LENGTH = NAME_LENGTH - 1; // leave space for null char
+	// leave space for leading/trailing '&' for container multidimensionality
+	static constexpr int NAME_MAX_LENGTH = VALUE_MAX_LENGTH - 2;
+
+	SCP_string container_name;
+	int type = SEXP_CONTAINER_LIST | SEXP_CONTAINER_STRING_DATA;
+	int opf_type = OPF_ANYTHING;
+	SCP_list<SCP_string> list_data;
+	SCP_unordered_map<SCP_string, SCP_string> map_data;
+
+	inline bool is_list() const
+	{
+		return type & SEXP_CONTAINER_LIST;
+	}
+
+	inline bool is_map() const
+	{
+		return type & SEXP_CONTAINER_MAP;
+	}
+
+	inline bool is_eternal() const
+	{
+		return type & SEXP_CONTAINER_SAVE_TO_PLAYER_FILE;
+	}
+
+	inline bool is_persistent() const
+	{
+		return type & SEXP_CONTAINER_IS_PERSISTENT;
+	}
+
+	bool empty() const
+	{
+		return (is_list() && list_data.empty()) || (is_map() && map_data.empty());
+	}
+
+	int size() const
+	{
+		if (is_list()) {
+			return (int)list_data.size();
+		} else if (is_map()) {
+			return (int)map_data.size();
+		} else {
+			Assert(false);
```

Comment:
Use `UNREACHABLE` here.

Reply:
Done. Glad there's a straightforward solution to this issue, because it pops up all over this place with this feature.

As a note, if we ever add a new container type (unlikely but possible), this block will become reachable if we forget to handle that new type in some spot, but then the code would need to be fixed, anyway.